orchestrator/orchestrator.go

```diff
@@ -0,0 +1,213 @@
+		}
+	}
+	for _, opSimInstance := range o.OpSimInstances {
+		if err := opSimInstance.Start(ctx); err != nil {
+			return fmt.Errorf("op simulator instance chain.id=%v failed to start: %w", opSimInstance.ChainId(), err)
+		}
+	}
+
+	if err := o.WaitUntilReady(); err != nil {
+		return fmt.Errorf("orchestrator failed to get ready: %w", err)
+	}
+
+	o.enableAnvilLogging()
+
+	o.log.Info("orchestrator is ready")
+
+	return nil
+}
+
+func (o *Orchestrator) Stop(ctx context.Context) error {
+	o.log.Info("stopping orchestrator")
+
+	for _, opSim := range o.OpSimInstances {
+		if err := opSim.Stop(ctx); err != nil {
+			return fmt.Errorf("op simulator chain.id=%v failed to stop: %w", opSim.ChainId(), err)
+		}
+		o.log.Info("stopped op simulator", "chain.id", opSim.ChainId())
+	}
+	for _, anvil := range o.anvilInstances {
+		if err := anvil.Stop(); err != nil {
+			return fmt.Errorf("anvil chain.id=%v failed to stop: %w", anvil.ChainId(), err)
+		}
+		o.log.Info("stopped anvil", "chain.id", anvil.ChainId())
+	}
+
+	o.log.Info("stopped orchestrator")
+
+	return nil
+}
+
+func (o *Orchestrator) Stopped() bool {
+	for _, anvil := range o.anvilInstances {
+		if stopped := anvil.Stopped(); !stopped {
+			return stopped
+		}
+	}
+	for _, opSim := range o.OpSimInstances {
+		if stopped := opSim.Stopped(); !stopped {
+			return stopped
+		}
+	}
+
+	return true
+}
+
+func (o *Orchestrator) WaitUntilReady() error {
+	var once sync.Once
+	var err error
+	ctx, cancel := context.WithCancel(context.Background())
+
+	handleErr := func(e error) {
+		if e != nil {
+			once.Do(func() {
+				err = e
+				cancel()
+			})
+		}
+	}
+
+	var wg sync.WaitGroup
+
+	waitForAnvil := func(anvil *anvil.Anvil) {
+		defer wg.Done()
+		handleErr(anvil.WaitUntilReady(ctx))
+	}
+
+	o.iterateAnvilInstances(func(chain *anvil.Anvil) {
+		wg.Add(1)
+		go waitForAnvil(chain)
+	})
+
+	wg.Wait()
+
+	return err
+}
+
+func (o *Orchestrator) iterateAnvilInstances(fn func(anvil *anvil.Anvil)) {
+	for _, anvilInstance := range o.anvilInstances {
+		fn(anvilInstance)
+	}
+}
+
+func (o *Orchestrator) enableAnvilLogging() {
+	o.iterateAnvilInstances(func(anvil *anvil.Anvil) {
+		anvil.EnableLogging()
+	})
+}
+
+func (o *Orchestrator) L1OpSimulator() *op_simulator.OpSimulator {
+	var result *op_simulator.OpSimulator
+	for _, opSim := range o.OpSimInstances {
+		if opSim.BaseChainId() == 0 {
+			result = opSim
+		}
+	}
+
+	return result
+}
+
+func (o *Orchestrator) L2OpSimulators() []*op_simulator.OpSimulator {
+	var result []*op_simulator.OpSimulator
+	for _, opSim := range o.OpSimInstances {
+		if opSim.BaseChainId() != 0 {
+			result = append(result, opSim)
+		}
+	}
+	return result
+}
+
+func (o *Orchestrator) ConfigAsString() string {
+	var b strings.Builder
+
+	fmt.Fprintf(&b, "\nSupersim Config:\n")
+	if o.L1OpSimulator() != nil {
+		fmt.Fprintf(&b, "L1:\n")
+		fmt.Fprintf(&b, "  Chain ID: %d    RPC: %s\n", o.L1OpSimulator().ChainId(), o.L1OpSimulator().Endpoint())
```

Comment:
we could consider encapsulating these strings if we wanted so it can just be something like `o.L1OpSimulator().String()`